File: ios/streaming/HlsPrefetchDeduper.swift
import Foundation

final class HlsPrefetchDeduper {
  private let window: TimeInterval
  private let maxEntries: Int
  private let now: () -> Date
  private let queue = DispatchQueue(label: "com.nitroplay.hls.prefetch-deduper")
  private var timestamps: [String: Date] = [:]
  private var insertionOrder: [String] = []

  init(window: TimeInterval, maxEntries: Int, now: @escaping () -> Date = Date.init) {
    self.window = window
    self.maxEntries = maxEntries
    self.now = now
  }

  var size: Int {
    queue.sync { timestamps.count }
  }

  func shouldPrefetch(key: String) -> Bool {
    queue.sync {
      let current = now()
      if let last = timestamps[key], current.timeIntervalSince(last) < window {
        return false
      }

      if timestamps[key] != nil {
        insertionOrder.removeAll { $0 == key }
      }
      insertionOrder.append(key)
      timestamps[key] = current
      trim(now: current)
      return true
    }
  }

  func forget(key: String) {
    queue.sync {
      timestamps.removeValue(forKey: key)
      insertionOrder.removeAll { $0 == key }
    }
  }

  func clear() {
    queue.sync {
      timestamps.removeAll()
      insertionOrder.removeAll()
    }
  }

  private func trim(now: Date) {
    guard timestamps.count > maxEntries else {
      return
    }

    let staleKeys = timestamps.compactMap { key, value in
      now.timeIntervalSince(value) > window ? key : nil
    }
    for key in staleKeys {
      timestamps.removeValue(forKey: key)
    }
    if !staleKeys.isEmpty {
      let stale = Set(staleKeys)
      insertionOrder.removeAll { stale.contains($0) }
    }

    var overBudget = timestamps.count - maxEntries
    guard overBudget > 0 else {
      return
    }

    while overBudget > 0, !insertionOrder.isEmpty {
      let oldest = insertionOrder.removeFirst()
      if timestamps.removeValue(forKey: oldest) != nil {
        overBudget -= 1
      }
    }
  }
}

File: android/src/main/java/com/nitroplay/streaming/HlsPrefetchDeduper.kt
package com.nitroplay.video.streaming

internal class HlsPrefetchDeduper(
  private val windowMs: Long,
  private val maxEntries: Int,
  private val nowMs: () -> Long = System::currentTimeMillis
) {
  private val timestamps = LinkedHashMap<String, Long>()
  private val lock = Any()

  val size: Int
    get() = synchronized(lock) { timestamps.size }

  fun shouldPrefetch(key: String): Boolean {
    synchronized(lock) {
      val now = nowMs()
      val last = timestamps[key]
      if (last != null && now - last < windowMs) {
        return false
      }

      timestamps.remove(key)
      timestamps[key] = now
      trim(now)
      return true
    }
  }

  fun forget(key: String) {
    synchronized(lock) {
      timestamps.remove(key)
    }
  }

  fun clear() {
    synchronized(lock) {
      timestamps.clear()
    }
  }

  private fun trim(now: Long) {
    if (timestamps.size <= maxEntries) return

    val stale = timestamps.entries.iterator()
    while (stale.hasNext()) {
      if (now - stale.next().value > windowMs) {
        stale.remove()
      }
    }

    var overBudget = timestamps.size - maxEntries
    if (overBudget <= 0) return

    val oldest = timestamps.entries.iterator()
    while (oldest.hasNext() && overBudget > 0) {
      oldest.next()
      oldest.remove()
      overBudget -= 1
    }
  }
}
